### PR TITLE
Added new operator S3FileTransformOperator

### PR DESCRIPTION
**Pull request by krishnabhupatiraju**
@mistercrunch @artwr Added a new S3FileTransformOperator operator. Please comment.

### CODE REVIEW

File: airflow/operators/s3_file_transform_operator.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            source_s3_conn_id='s3_default',
+            dest_s3_conn_id='s3_default',
+            replace_destination=False,
+            *args, **kwargs):
+        super(S3FileTransformOperator, self).__init__(*args, **kwargs)
+        self.source_s3_key = source_s3_key
+        self.source_s3_conn_id = source_s3_conn_id
+        self.dest_s3_key = dest_s3_key
+        self.dest_s3_conn_id = dest_s3_conn_id
+        self.replace_destination = replace_destination
```
Comment (artwr) [inline airflow/operators/s3_file_transform_operator.py]: Maybe just "replace", or "overwrite"

File: airflow/operators/s3_file_transform_operator.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            source_s3_key_object.get_contents_to_file(f_source)
+            f_source.flush()
+            self.source_s3.connection.close()
+            transform_script_process = subprocess.Popen([self.transform_executor, self.transform_script, f_source.name, f_dest.name], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            (transform_script_stdoutdata, transform_script_stderrdata) = transform_script_process.communicate()
+            logging.info("Transform script stdout " + transform_script_stdoutdata)
+            if transform_script_process.returncode > 0:
+                raise Exception("Transform script failed " + transform_script_stderrdata)
+            else:
+                logging.info("Transform script successful. Output located at {0}".format(f_dest.name))
```
Comment (artwr) [inline airflow/operators/s3_file_transform_operator.py]: Unfortunately or fortunately, with the clean up facilities in place in tempfile, this file should never be available.

File: airflow/operators/s3_file_transform_operator.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            logging.info("Uploading transformed file to S3")
+            self.dest_s3.load_file(
+                filename=f_dest.name, 
+                key=self.dest_s3_key, 
+                replace=self.replace_destination
+                )
+            f_dest.flush()
```
Comment (artwr) [inline airflow/operators/s3_file_transform_operator.py]: Not needed. This is to flush to disk

File: airflow/operators/s3_file_transform_operator.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            self.dest_s3.connection.close()
+
```
Comment (artwr) [inline airflow/operators/s3_file_transform_operator.py]: Maybe log success after that?
logging.info("Upload succeeded")

**Comment by artwr** [comment]
Overall ok, minor nits.
I am not completely convinced that this is an operator with much generality. Should we move this to contrib in the data repo? @mistercrunch ?

File: airflow/operators/s3_file_transform_operator.py (hunk 2 above)
Comment (krishnabhupatiraju) [inline airflow/operators/s3_file_transform_operator.py]: This was really just for logging than anything else. Ill change the message.

File: airflow/operators/s3_file_transform_operator.py (hunk 1 above)
Comment (krishnabhupatiraju) [inline airflow/operators/s3_file_transform_operator.py]: Makes sense. Changed

File: airflow/operators/s3_file_transform_operator.py (hunk 3 above)
Comment (krishnabhupatiraju) [inline airflow/operators/s3_file_transform_operator.py]: Moved this to the previous line so contents are flushed to disk before it is loaded to S3.

File: airflow/operators/s3_file_transform_operator.py (hunk 4 above)
Comment (krishnabhupatiraju) [inline airflow/operators/s3_file_transform_operator.py]: Added.

File: airflow/operators/s3_file_transform_operator.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    :type source_s3_key: str
+    :param source_s3_conn_id: source s3 connection
+    :type source_s3_conn_id: str
+    :param dest_s3_key: The key to be written from S3
+    :type dest_s3_key: str
+    :param dest_s3_conn_id: destination s3 connection
+    :type dest_s3_conn_id: str
+    :param replace: Replace dest S3 key if it already exists
+    :type replace: bool
+    :param transform_executor: location of the executor of the transformation script
```
Comment (mistercrunch) [inline airflow/operators/s3_file_transform_operator.py]: probably not needed, let's assume an executable for the `transform_script` param

File: airflow/operators/s3_file_transform_operator.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                raise Exception("Transform script failed " + transform_script_stderrdata)
+            else:
+                logging.info("Transform script successful. Output temporarily located at {0}".format(f_dest.name))
+            logging.info("Uploading transformed file to S3")
+            f_dest.flush()
+            self.dest_s3.load_file(
+                filename=f_dest.name, 
+                key=self.dest_s3_key, 
+                replace=self.replace
+                )
```
Comment (mistercrunch) [inline airflow/operators/s3_file_transform_operator.py]: dedent

**Comment by mistercrunch** [comment]
Looks good! feel free to merge. We'll have to get unit tests for these S3 hooks at some point.

**Comment by krishnabhupatiraju** [comment]
Made changes suggested by Artwr and Mistercrunch. Merging.

**Pull request merged**